src/OSPSuite.Infrastructure.Import/Core/DataSourceFileReaders/CsvDataSourceFile.cs

```diff
@@ -17,48 +16,52 @@ public CsvDataSourceFile(IImportLogger logger, ICsvSeparatorSelector csvSeparato
       {
          _csvSeparatorSelector = csvSeparatorSelector;
       }
-      protected override Cache<string, DataSheet> LoadFromFile(string path)
+      protected override void LoadFromFile(string path)
       {
+         var separator = _csvSeparatorSelector.GetCsvSeparator(path);
+
+         //if separator selection dialog was cancelled, abort
+         if (!(separator is char separatorCharacter)) return;
+
+         //we keep a copy of the already loaded sheets, in case the reading fails
+         var alreadyLoadedDataSheets = new DataSheetCollection(DataSheets);
+         DataSheets.Clear();
+
          try
          {
-            var separator = _csvSeparatorSelector.GetCsvSeparator(path);
-
-            //if separator selection dialog was cancelled, abort
-            if (!(separator is char separatorCharacter)) return null;
- 
             using (var reader = new CsvReaderDisposer(path, separatorCharacter))
             {
                var csv = reader.Csv;
                var headers = csv.GetFieldHeaders();
                var rows = new List<List<string>>(headers.Length);
 
-               var dataSheet = new DataSheet {RawData = new UnformattedData()};
+               var dataSheet = new DataSheet
+               {
+                  SheetName = ""
+               };
 
                for (var i = 0; i < headers.Length; i++)
-                  dataSheet.RawData.AddColumn(headers[i], i);
+                  dataSheet.AddColumn(headers[i], i);
                var currentRow = new string[csv.FieldCount];
 
                while (csv.ReadNextRecord())
                {
                   csv.CopyCurrentRecordTo(currentRow);
                   var rowList = currentRow.ToList();
                   var levels = getMeasurementLevels(rowList);
-                  dataSheet.RawData.CalculateColumnDescription(levels);
-                  dataSheet.RawData.AddRow(rowList);
+                  dataSheet.CalculateColumnDescription(levels);
+                  dataSheet.AddRow(rowList);
                }
 
-               dataSheet.RawData.RemoveEmptyColumns();
-               dataSheet.RawData.RemoveEmptyRows();
+               dataSheet.RemoveEmptyColumns();
+               dataSheet.RemoveEmptyRows();
 
-               var loadedData = new Cache<string, DataSheet>()
-               {
-                  { "", dataSheet }
-               };
-               return loadedData;
+               DataSheets.AddSheet(dataSheet);
             }
          }
          catch (Exception e)
          {
+            DataSheets = alreadyLoadedDataSheets;
```

Comment:
Shouldn't DataSheet ( the member ) be read-only?

Reply:
I removed the setter from the DataSheet property